### ISSUE TITLE
pci: vfio: support IOMMUFD API

### ISSUE THREAD

**Issue by edigaryev** [cloud-hypervisor/cloud-hypervisor#6892]
Currently, it's possible to benefit from the advanced featured offered by the IOMMUFD through a kernel compatibility layer, which can enabled using `CONFIG_IOMMUFD_VFIO_CONTAINER` kernel option, however, it lacks some features<sup>1</sup>:

>Note that at the time of writing, the compatibility mode is not entirely feature complete relative to VFIO_TYPE1v2_IOMMU (ex. DMA mapping MMIO) and does not attempt to provide compatibility to the VFIO_SPAPR_TCE_IOMMU interface. Therefore it is not generally advisable at this time to switch from native VFIO implementations to the IOMMUFD compatibility interfaces.

It's recommended to use IOMMUFD API directly in the long-run<sup>1</sup>:

>Long term, VFIO users should migrate to device access through the cdev interface described below, and native access through the IOMMUFD provided interfaces.

See also QEMU's [IOMMUFD BACKEND usage with VFIO](https://www.qemu.org/docs/master/devel/vfio-iommufd.html).

<sup>1</sup>: https://docs.kernel.org/driver-api/vfio.html#iommufd-and-vfio-iommu-type1

**Comment by rbradford** [comment]
Is there a call to action here?

**Comment by edigaryev** [comment]
> Is there a call to action here?

Yes, it would be nice to see native IOMMUFD support in Cloud Hypervisor.

**Comment by rbradford** [comment]
Gotcha - thanks! The vfio crate needs support first - https://github.com/rust-vmm/vfio
